prometheus/push.go

```diff
@@ -19,47 +19,153 @@
+				}
+				if _, ok := grouping[l.GetName()]; ok {
+					return fmt.Errorf(
+						"pushed metric %s (%s) already contains grouping label %s",
+						mf.GetName(), m, l.GetName(),
+					)
+				}
+			}
+		}
+		enc.Encode(mf)
+	}
+	req, err := http.NewRequest(method, pushURL, buf)
+	if err != nil {
+		return err
+	}
+	req.Header.Set(contentTypeHeader, string(expfmt.FmtProtoDelim))
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return err
+	}
+	defer resp.Body.Close()
+	if resp.StatusCode != 202 {
+		body, _ := ioutil.ReadAll(resp.Body) // Ignore any further error as this is for an error message only.
+		return fmt.Errorf("unexpected status code %d while pushing to %s: %s", resp.StatusCode, pushURL, body)
+	}
+	return nil
 }
 
-// PushCollectors works like Push, but it does not collect from the default
-// registry. Instead, it collects from the provided collectors. It is a
+// PushCollectors works like PushFrom, but it does not collect via a
+// registry. Instead, it collects from the provided collectors directly. It is a
 // convenient way to push only a few metrics.
-func PushCollectors(job, instance, url string, collectors ...Collector) error {
-	return pushCollectors(job, instance, url, "PUT", collectors...)
+func PushCollectors(job string, grouping map[string]string, url string, collectors ...Collector) error {
+	return pushCollectors(job, grouping, url, "PUT", collectors...)
 }
 
-// PushAddCollectors works like PushAdd, but it does not collect from the
-// default registry. Instead, it collects from the provided collectors. It is a
+// PushAddCollectors works like PushAddFrom, but it does not collect via a
+// registry. Instead, it collects from the provided collectors directly. It is a
 // convenient way to push only a few metrics.
-func PushAddCollectors(job, instance, url string, collectors ...Collector) error {
-	return pushCollectors(job, instance, url, "POST", collectors...)
+func PushAddCollectors(job string, grouping map[string]string, url string, collectors ...Collector) error {
+	return pushCollectors(job, grouping, url, "POST", collectors...)
 }
 
-func pushCollectors(job, instance, url, method string, collectors ...Collector) error {
-	r := newRegistry()
+func pushCollectors(job string, grouping map[string]string, url, method string, collectors ...Collector) error {
+	r := NewRegistry()
 	for _, collector := range collectors {
-		if _, err := r.Register(collector); err != nil {
+		if err := r.Register(collector); err != nil {
 			return err
 		}
 	}
-	return r.Push(job, instance, url, method)
+	return push(r, job, grouping, url, method)
+}
+
+// GroupByHostname returns a label map with the only entry
+// {instance="<hostname>"}. This can be conveniently used as the grouping
+// parameter if metrics should be pushed with the hostname as label. The
+// returned map is created upon each call so that the caller is free to add more
+// labels to the map.
+func GroupByHostname() map[string]string {
```

Comment:
How have you picked the IP number in the other clients? The first one that doesn't look like localhost?